### PR TITLE
Fixes #1678: Optimize Bazel protobuf rules install [BLOCKED: #1634]

### LINKED ISSUE #1678: Proto rules in WORKSPACE install unnecessary compilers
# Problem
Currently, the following are used to import the external dependencies needed to run proto rules like `proto_library` and `java_lite_proto_library` in Bazel:

``` 
# rules_proto defines abstract rules for building Protocol Buffers.
http_archive(
    name = "rules_proto",
    sha256 = "2490dca4f249b8a9a3ab07bd1ba6eca085aaf8e45a734af92aad0c42d9dc7aaf",
    strip_prefix = "rules_proto-218ffa7dfa5408492dc86c01ee637614f8695c45",
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/rules_proto/archive/218ffa7dfa5408492dc86c01ee637614f8695c45.tar.gz",
        "https://github.com/bazelbuild/rules_proto/archive/218ffa7dfa5408492dc86c01ee637614f8695c45.tar.gz",
    ],
)

load("@rules_proto//proto:repositories.bzl", "rules_proto_dependencies", "rules_proto_toolchains")
rules_proto_dependencies()
rules_proto_toolchains()

# rules_java defines rules for generating Java code from Protocol Buffers.
http_archive(
    name = "rules_java",
    sha256 = "ccf00372878d141f7d5568cedc4c42ad4811ba367ea3e26bc7c43445bbc52895",
    strip_prefix = "rules_java-d7bf804c8731edd232cb061cb2a9fe003a85d8ee",
    urls = [
        "https://mirror.bazel.build/github.com/bazelbuild/rules_java/archive/d7bf804c8731edd232cb061cb2a9fe003a85d8ee.tar.gz",
        "https://github.com/bazelbuild/rules_java/archive/d7bf804c8731edd232cb061cb2a9fe003a85d8ee.tar.gz",
    ],
)

load("@rules_java//java:repositories.bzl", "rules_java_dependencies", "rules_java_toolchains")
rules_java_dependencies()
rules_java_toolchains()
```

However, these `http_archive` rules bring in more than what is used in Oppia Android. Some of these downloads are lengthly and expensive - so optimizing this install is imperative. 

# Solution

Thanks to @anandwana001, the following solution was found: https://github.com/google/startup-os/blob/master/WORKSPACE#L173. Replacing the `http_archive` rules above with these bindings and removing the now unecessary `load()` statements from `model/BUILD` fixes this issue.

### PR DESCRIPTION
**Pull request by miaboloix**
## Explanation
Fixes #1678: Optimize Bazel protobuf rules install.

Currently, `http_rules` in WORKSPACE are used to import `rules_java` and `rules_proto`, the external dependencies needed to run proto rules like `proto_library` and `java_lite_proto_library` in Bazel. 

However, these `http_archive` rules bring in more than what is needed to use  `proto_library` and `java_lite_proto_library` and some of these unused artifacts come with lengthly and expensive download times.

Thanks to @anandwana001, the following solution was found: https://github.com/google/startup-os/blob/master/WORKSPACE#L173. By replacing the `http_archive` rules with these bindings and removing the now unnecessary `load()` statements from `model/BUILD`, we are able to import only what is necessary. 

Building app library from clean...
- Before optimization: ~3min 45s
- After optimization: ~2min 55s

## Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [ ] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR is made from a branch that is up-to-date with "develop".
- [ ] The PR's branch is based on "develop" and not on any other branch.
- [ ] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Review by anandwana001 — APPROVED** [review]
LGTM, Thanks @miaboloix 
Yes, I think this approach is a good to go for.

File: WORKSPACE [hunk 1]
```diff
@@ -43,30 +43,19 @@ load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kotlin_repositories", "kt_reg
...
-# rules_java defines rules for generating Java code from Protocol Buffers.
-http_archive(
-    name = "rules_java",
-    sha256 = "ccf00372878d141f7d5568cedc4c42ad4811ba367ea3e26bc7c43445bbc52895",
-    strip_prefix = "rules_java-d7bf804c8731edd232cb061cb2a9fe003a85d8ee",
-    urls = [
-        "https://mirror.bazel.build/github.com/bazelbuild/rules_java/archive/d7bf804c8731edd232cb061cb2a9fe003a85d8ee.tar.gz",
-        "https://github.com/bazelbuild/rules_java/archive/d7bf804c8731edd232cb061cb2a9fe003a85d8ee.tar.gz",
-    ],
+bind(
```
Comment (BenHenning) [inline WORKSPACE:57]: Is this using the protobuf bundled with Bazel? Is this the recommended way to do this? Can you point here & in code to documentation that explains this?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @miaboloix! Can you include in the PR description before & after build times to quantify what sort of improvement we get with this optimization?

File: WORKSPACE (hunk 1 above)
Comment (miaboloix) [reply WORKSPACE:57]: Sure - there aren't any in-depth explanations but I have linked documentation related to the bind() command, where this technique is used, and an issue discussing something similar

File: WORKSPACE [hunk 2]
```diff
@@ -43,30 +43,21 @@ load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kotlin_repositories", "kt_reg
...
-    urls = [
-        "https://mirror.bazel.build/github.com/bazelbuild/rules_proto/archive/218ffa7dfa5408492dc86c01ee637614f8695c45.tar.gz",
-        "https://github.com/bazelbuild/rules_proto/archive/218ffa7dfa5408492dc86c01ee637614f8695c45.tar.gz",
-    ],
+'''
+The proto_compiler and proto_java_toolchain bindings load the protos rules needed for the model
+module while helping us avoid the unnecessary compilation of protoc. Referecences:
+https://docs.bazel.build/versions/master/be/workspace.html#bind,
+https://github.com/google/startup-os/blob/master/WORKSPACE#L215-L220,
+https://github.com/bazelbuild/bazel/issues/7095
```
Comment (BenHenning) [inline WORKSPACE]:
```suggestion
- https://docs.bazel.build/versions/master/be/workspace.html#bind
- https://github.com/google/startup-os/blob/master/WORKSPACE#L215-L220
- https://github.com/bazelbuild/bazel/issues/7095
```

Markdown is a nice way to enumerate items in a readable way.

File: WORKSPACE [hunk 3]
```diff
@@ -43,30 +43,21 @@ load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kotlin_repositories", "kt_reg
...
-    strip_prefix = "rules_proto-218ffa7dfa5408492dc86c01ee637614f8695c45",
-    urls = [
-        "https://mirror.bazel.build/github.com/bazelbuild/rules_proto/archive/218ffa7dfa5408492dc86c01ee637614f8695c45.tar.gz",
-        "https://github.com/bazelbuild/rules_proto/archive/218ffa7dfa5408492dc86c01ee637614f8695c45.tar.gz",
-    ],
+'''
+The proto_compiler and proto_java_toolchain bindings load the protos rules needed for the model
+module while helping us avoid the unnecessary compilation of protoc. Referecences:
+https://docs.bazel.build/versions/master/be/workspace.html#bind,
+https://github.com/google/startup-os/blob/master/WORKSPACE#L215-L220,
```
Comment (BenHenning) [inline WORKSPACE]: Is this pointing to the right thing? It seems unrelated. Also, suggest including commit hashes instead of 'blob' to permalink changes (otherwise files change over time since 'blob' corresponds to tip-of-tree). Suggest using the short 7 character hash to keep the line length short.

File: WORKSPACE (hunk 1 above)
Comment (BenHenning) [reply WORKSPACE:57]: Thanks for the documentation. It slightly confuses me, though. The Bazel issue you linked to (https://github.com/bazelbuild/bazel/issues/7095) does discuss this as a problem, but the Bazel documentation actually suggests against using bind and that earlier issue doesn't mention bind as a solution.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Following back up. Fantastic build time improvement! I still want to better understand whether this is the correct approach & why, and ensure that's documented before we go ahead with the solution.

File: WORKSPACE (hunk 1 above)
Comment (miaboloix) [reply WORKSPACE:57]: Yes, I couldn't find anything directly explaining this solution. I just have the example displayed here: https://github.com/google/startup-os/blob/master/WORKSPACE#L173 and found the thread I linked. In terms of the documentation for bind(), I figured it could be helpful if developers want to look into what the function does? But I guess they could also just google it?

Comment (miaboloix) [reply WORKSPACE:58]: okay!

File: WORKSPACE (hunk 2 above)
Comment (miaboloix) [reply WORKSPACE]: Done!

File: WORKSPACE (hunk 3 above)
Comment (miaboloix) [reply WORKSPACE]: Fixed!

File: WORKSPACE (hunk 1 above)
Comment (BenHenning) [reply WORKSPACE:57]: Okay, makes sense. Thanks!

**Review by BenHenning — APPROVED** [review]
Thanks @miaboloix!
